### ISSUE TITLE
"format specifies type" compilation warnings with clang

### ISSUE THREAD

**Issue by anatoli26** [cyrusimap/cyrus-imapd#3128]
When compiling cyrus-imapd with llvm/clang on OpenBSD, it generates about 60 "format specifies type" warnings, like:

```
imap/http_carddav.c:966:39: warning: format specifies type 'long' but the argument has type 'time_t'
(aka 'long long') [-Wformat]
    buf_printf(&txn->buf, "-%ld-%ld", sbuf.st_mtime, sbuf.st_size);
                            ~~~       ^~~~~~~~~~~~~
                            %lld
```

This is due to the types like `off_t` and `time_t` being defined as `long long` on OpenBSD and on Linux they are defined as `long int`, which in practice on amd64 is actually the same, but not precise enough for clang.

AFAIK, the most portable way to solve this is for the compilers supporting C99 to use the following `printf`/`scanf` length modifiers:
* `z` for `ssize_t`,
* `t` for `ptrdiff_t`,
* `q` for `quad_t`, and
* `j` for everything else (it's a length modifier for `intmax_t` - the largest `int` supported on the platform).

For example:

```
buf_printf(&txn->buf, "-%ld-%ld", sbuf.st_mtime, sbuf.st_size);
```

becomes

```
buf_printf(&txn->buf, "-%jd-%jd", sbuf.st_mtime, sbuf.st_size);
```

or even:

```
buf_printf(&txn->buf, "-%jd-%jd", (intmax_t)sbuf.st_mtime, (intmax_t)sbuf.st_size);
```

if you want a bullet-proof implementation for the foreseeable future (the cast is needed if the type to print is smaller than `intmax_t`).

The complete list of warnings is as follows:

```
imap/http_admin.c:259:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_admin.c:259:31: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_admin.c:605:44: warning: format specifies type 'long *' but the argument has type 'time_t *' (aka 'long long *') [-Wformat]
imap/http_admin.c:797:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_admin.c:797:31: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_admin.c:1129:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_admin.c:1129:31: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_caldav_sched.c:352:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_carddav.c:961:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_carddav.c:961:30: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_carddav.c:961:45: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_carddav.c:966:39: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_carddav.c:966:54: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_caldav.c:2268:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_caldav.c:2268:30: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_caldav.c:2268:45: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_caldav.c:2273:39: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_caldav.c:2273:54: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_dav.c:1474:52: warning: format specifies type 'unsigned long' but the argument has type 'long long' [-Wformat]
imap/http_h2.c:327:37: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_h2.c:743:37: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_h2.c:881:37: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_dav.c:8599:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_dav.c:8599:31: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_ischedule.c:227:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_ischedule.c:227:31: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_ischedule.c:272:38: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_ischedule.c:664:26: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_dav_sharing.c:1142:52: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_rss.c:616:39: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_rss.c:616:54: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_rss.c:621:39: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_rss.c:621:54: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/http_ws.c:314:44: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_ws.c:339:44: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_tzdist.c:1036:47: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_tzdist.c:1143:45: warning: format specifies type 'long *' but the argument has type 'time_t *' (aka 'long long *') [-Wformat]
imap/http_tzdist.c:1154:55: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_tzdist.c:1833:52: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/http_tzdist.c:2040:52: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/httpd.c:2685:62: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/httpd.c:3762:36: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/httpd.c:4398:16: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/httpd.c:4398:30: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/httpd.c:4398:45: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]
imap/zoneinfo_db.c:191:68: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
sieve/sieved.c:161:34: warning: format specifies type 'unsigned long' but the argument has type 'unsigned long long' [-Wformat]
sieve/sieved.c:161:44: warning: format specifies type 'unsigned long' but the argument has type 'unsigned long long' [-Wformat]
sieve/sieved.c:161:61: warning: format specifies type 'unsigned long' but the argument has type 'unsigned long long' [-Wformat]
sieve/sieved.c:176:29: warning: format specifies type 'unsigned long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
imap/caldav_alarm.c:458:32: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/caldav_alarm.c:609:43: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/caldav_alarm.c:609:58: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/caldav_alarm.c:639:16: warning: format specifies type 'long *' but the argument has type 'time_t *' (aka 'long long *') [-Wformat]
imap/caldav_alarm.c:639:32: warning: format specifies type 'long *' but the argument has type 'time_t *' (aka 'long long *') [-Wformat]
imap/caldav_alarm.c:458:32: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/caldav_alarm.c:609:43: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/caldav_alarm.c:609:58: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
imap/caldav_alarm.c:639:16: warning: format specifies type 'long *' but the argument has type 'time_t *' (aka 'long long *') [-Wformat]
imap/caldav_alarm.c:639:32: warning: format specifies type 'long *' but the argument has type 'time_t *' (aka 'long long *') [-Wformat]
```

**Comment by obache** [comment]
10 years, long standing no feedback issue #1287, how format time_t.

**Comment by elliefm** [comment]
The right way to do this will be similar to how we handle SIZE_T_FMT, OFF_T_FMT, and maybe others:  detect and #define TIME_T_FMT to be the right thing during configure, and then use TIME_T_FMT everywhere instead of whatever we're using at the moment.

The C99 z/j/etc modifiers are, annoyingly, not portable enough [insert obligatory xkcd 927 reference], and this whole thing ends up being a nuisance for pretty much every portable C project ever.  These days, there is probably a pre-existing set of autoconf macros for handling it... but I believe (without looking) that what we're using for SIZE_T_FMT and OFF_T_FMT is probably homegrown, because they're defined differently from the standard PRI macros from inttypes.h (n.b. the standard set do not include the '%', but our ones do).

It sounds like there's probably also a bunch of places that _should_ be using SIZE_T_FMT or OFF_T_FMT but presently aren't, and that they need fixing too.

I could see an argument in favour of switching to whatever macros autoconf provides, and standardising on those, but that seems like a big and tedious task.

**Comment by anatoli26** [comment]
We could use `__STDC_VERSION__` to detect C99 and proceed with `%z`/`%j` & `intmax_t` cast if we have support, and with `%lld` if we don't. But in that case why not just to convert every occurrence to `%lld`/`%llu` as corresponds (as was [suggested](https://github.com/cyrusimap/cyrus-imapd/issues/1287#issuecomment-266649669) by Antoine Jacoutot (the cyrus-imapd port author at OpenBSD) in the related issue #1287)?

I could submit a PR with the changes fixing all the mismatches of the format specifiers as suggested by clang, which are actually more than 60: the 60 are the warnings reported currently during build of the OpenBSD port + 10 patches with 3-4 changes each that are fixing some of the mismatches. So it should be around 100 fixes needed.

**Comment by elliefm** [comment]
I think all the "off_t" warnings should be corrected by just changing them to use the existing OFF_T_FMT.  

For example this one:

>imap/http_admin.c:259:31: warning: format specifies type 'long' but the argument has type 'off_t' (aka 'long long') [-Wformat]

is this:
```
    buf_printf(&txn->buf, "%ld-%ld-%ld", (long) compile_time,
               sbuf.st_mtime, sbuf.st_size);
```

Note that it's using hardcoded "%ld" for sbuf.st_size, when it should be using OFF_T_FMT for that one.

Changing it to something like this (untested, written straight into the github comment box):
```
    buf_printf(&txn->buf, "%ld-%ld-" OFF_T_FMT, (long) compile_time,
               sbuf.st_mtime, sbuf.st_size);
```

would correct that warning (though it would still warn about the bad time_t in the same line, because this trivial example only fixes the off_t).

We can add a similar thing that detects the size of time_t and defines an appropriate TIME_T_FMT, and then use that for all the time_t ones.

Casting in this situation is a bad idea.  From a maintenance point of view, casting usually means "there's something tricky going on here; tread carefully".  If we start littering every formatted string with casts, it becomes impossible to tell which ones are actually something tricky, and which ones are just annoying platform workarounds.  The right place for platform workarounds is configure.ac, and we already have half of this in place; it's just a matter of finishing the job.

If we had a TIME_T_FMT established by configure.ac to match the others, then the example fix above would become something like:

```
    buf_printf(&txn->buf, "%ld-" TIME_T_FMT "-" OFF_T_FMT, (long) compile_time,
               sbuf.st_mtime, sbuf.st_size);
```

and that would resolve both of the warnings from this line.

**Comment by anatoli26** [comment]
Sounds good. I'll prepare a patch to create all the missing XX_T_FMT and then start submitting changes for the printf/scanf.

**Comment by obache** [comment]
casting `compile_time` to `long` should be fixed too.

**Comment by anatoli26** [comment]
@elliefm,

Please check these warnings:

```
imap/http_dav.c:1474:52: warning: format specifies type 'unsigned long' but the argument has type 'long long' [-Wformat]
        snprintf(tbuf, sizeof(tbuf), "Second-%lu", data->lock_expire - now);
                                             ~~~   ^~~~~~~~~~~~~~~~~~~~~~~

sieve/sieved.c:161:34: warning: format specifies type 'unsigned long' but the argument has type 'unsigned long long' [-Wformat]
    printf(" %02lu:%02lu:%02lu", t / 3600, (t % 3600) / 60, t % 60);
             ~~~~~               ^~~~~~~~
--
sieve/sieved.c:161:44: warning: format specifies type 'unsigned long' but the argument has type 'unsigned long long' [-Wformat]
    printf(" %02lu:%02lu:%02lu", t / 3600, (t % 3600) / 60, t % 60);
                   ~~~~~                   ^~~~~~~~~~~~~~~
--
sieve/sieved.c:161:61: warning: format specifies type 'unsigned long' but the argument has type 'unsigned long long' [-Wformat]
    printf(" %02lu:%02lu:%02lu", t / 3600, (t % 3600) / 60, t % 60);
                         ~~~~~                              ^~~~~~
--
sieve/sieved.c:176:29: warning: format specifies type 'unsigned long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
        else printf(" %lu", i);
                      ~~~   ^
```

The first one, I guess, should use the `TIME_T_FMT`, but I'm not sure what does the `print_time` func do in the last 4 cases.

How should these pieces of code be modified?

**Comment by elliefm** [comment]
> The first one, I guess, should use the `TIME_T_FMT`

Yep, that one's just a straight `time_t` so `TIME_T_FMT` is the correct one to use.

The next three are the same line, and an annoyingly platform-specific case.  On my platform, `lu` is appropriate for `uint64_t`, but I suspect on your platform it wants `llu`.  The correct thing to do here is use the corresponding C99 inttypes.h macro, in this case `PRIu64`.  Something like (typed straight into comment box, untested):

```
    printf(" %02" PRIu64 ":%02" PRIu64 ":%02" PRIu64, t / 3600, (t % 3600) / 60, t % 60);
```

(This line will probably need to be rewrapped too; just follow the existing conventions.)

You also should `#include <inttypes.h>` near the top of this file for the PRIx macros to be defined.  It may be already included implicitly through something else, but it's better to be explicit.

The last one is also a `uint64_t`, so it needs the same treatment.  Something like:

```
        else printf(" %" PRIu64, i); 
```

**Comment by anatoli26** [comment]
Good, thanks for the details. I'll submit the PR now with the necessary changes.

**Comment by anatoli26** [comment]
@elliefm, I was testing the format specifiers like we discussed here and already converted half of the code, but then I noticed that the defines of this form:

```
#if (SIZEOF_TIME_T == SIZEOF_LONG)
#define TIME_T_FMT "%ld"
#define strtoofft(nptr, endptr, base) strtol(nptr, endptr, base)                                                                                                                                                                                                                                                              
#elif (SIZEOF_TIME_T == SIZEOF_LONG_LONG_INT)
#define TIME_T_FMT "%lld"
#define strtoofft(nptr, endptr, base) strtoll(nptr, endptr, base)
#else
#error dont know what to use for TIME_T_FMT
#endif
```

actually don't work as `SIZEOF_TIME_T` is 8 and both `SIZEOF_LONG` and `SIZEOF_LONG_LONG_INT` are also 8. So the first condition matches and we are back at where we started...

From `configure`:
```
checking size of int... 4                                                                                                                                                                                                                                                                                                     
checking size of long... 8                                                                                                                                                                                                                                                                                                    
checking size of size_t... 8                                                                                                                                                                                                                                                                                                  
checking size of off_t... 8                                                                                                                                                                                                                                                                                                   
checking size of time_t... 8                                                                                                                                                                                                                                                                                                  
checking size of long long int... 8                                                                                                                                                                                                                                                                                           
checking size of unsigned long long int... 8        
```

Any ideas how to distinguish the needed specifier?

**Comment by elliefm** [comment]
Wow that's annoying!

It _might_ be as simple as just reversing the order of those two checks (i.e. do the SIZEOF_LONG_LONG_INT check first, so that if both types are the same size it will prefer the "longer" format string).

If that doesn't just fix it, can you provide the compile error that occurs?  It might shed some more light, it might not...

**Comment by anatoli26** [comment]
If we reverse the order, we'd get `%ldd` on Linux too. I.e. these tests are not good to distinguish the base type of the typedef of these special types when the platform is amd64.

I was reading the autoconf docs. There is AC_CHECK_TYPE and related AC_TYPE_OFF_T & similar, but it only checks for the existence of the type, not its base type.

I don't know well autoconf, maybe there are some ways to get this info, but at the moment I don't know what could be done. We need to get the exact base type for each of the special types we're interested in to be able to proceed with the definition of the format specifiers.

**Comment by elliefm** [comment]
I don't think it would be a problem on Linux, but I'm not certain of that.  If reversing them works okay on OpenBSD,  I think stick with that and get the PR submitted, and then I'll test it on Linux here and see if anything complains.  And if anything does complain, we can look into more elaborate solutions once we know we need them ;)

**Comment by anatoli26** [comment]
I just created a PR to accommodate all the issues with format specifiers discussed here: #3262.

@obache, I've also fixed the `compile_time` casts (included in the [3rd commit](https://github.com/cyrusimap/cyrus-imapd/pull/3262/commits/8bb3293d12adcdd98cf414b9c4b8b535d1c55c23)).

@elliefm, please test it on different platforms (i386, amd64, arm, etc.) and with different OS as the `configure.ac` changes together with everything else could cause unexpected results.

**Comment by anatoli26** [comment]
>> If we reverse the order, we'd get %ldd on Linux too. I.e. these tests are not good to distinguish the base type of the typedef of these special types when the platform is amd64.
>> 
>> I was reading the autoconf docs. There is AC_CHECK_TYPE and related AC_TYPE_OFF_T & similar, but it only checks for the existence of the type, not its base type.
>> 
>> I don't know well autoconf, maybe there are some ways to get this info, but at the moment I don't know what could be done. We need to get the exact base type for each of the special types we're interested in to be able to proceed with the definition of the format specifiers.

> I don't think it would be a problem on Linux, but I'm not certain of that. If reversing them works okay on OpenBSD, I think stick with that and get the PR submitted, and then I'll test it on Linux here and see if anything complains. And if anything does complain, we can look into more elaborate solutions once we know we need them ;)

@elliefm, I asked for help on this issue from autoconf@gnu.org mailinglist, and here goes one suggestion (edited by me, the original is here [1] and the message that goes after it):

---

_You can use C11_ _**_Generic** and **AC_COMPUTE_INT** to pretty easily and quickly determine which type (out of a finite list of candidates) time_t or any other type is compatible with._

_But you'd need a fallback (probably by compiling one program like the one shown above for each type) to handle the case where_ __Generic is not supported by the implementation._

_Example (totally untested):_

```
  AC_COMPUTE_INT([timetype],
    [_Generic((time_t)0, long long: 3, long: 2, int: 1, default: 0)],
    [#include <time.h>],
    [... slow fallback computation goes here])

  AS_CASE([$timetype],
    [3], [... action when time_t is compatible with long long],
    [2], [... action when time_t is compatible with long],
    [1], [... action when time_t is compatible with int],
    [... action when time_t's compatibility is undetermined])
```
---

And if we could rely on C11, we could actually have a define in the code like this and not need any hacks with autoconf at all:

```
time_t _unused_t;
#define TIME_T_FMT _Generic((_unused_t), long long int: "%lld", long int: "%ld", int: "%d")
```

We could specify a `default` value or, if it's not specified and none of the options matches, it would cause a compile-time error, which could be a desired outcome so we detect any type issues immediately.

Also, we could have a generic `#define` that would determine the format specifier for any int var, e.g.:
```
#define _FTM(x) _Generic((x), long long int: "%lld", long int: "%ld", int: "%d", \
unsigned long long int: "%llu", unsigned long int: "%lu", unsigned int: "%u")
```
and use it in the printf & co like this:

```
buf_printf(&buf, _FMT(record->savedate), record->savedate);
syslog(LOG_NOTICE, "SYNCNOTICE: %s uid:" _FMT(record->uid) " modseq:" _FMT(record->modseq)
	" last_updated:" _FMT(record->last_updated) " internaldate:" _FMT(record->internaldate)
	" flags:(%s) cid:" _FMT(record->cid),
	name, record->uid, record->modseq,
	record->last_updated, record->internaldate,
	make_flags(mailbox, record), record->cid);
```
The added benefit would be that you won't have to find all the places where a var is printed when you change its type. You either always have the correct format specifier for the var, or you get a compilation error.

The question is can we restrict the compilation to C11 and newer?

And if we can't, we could use another technique (this time with autoconf) suggested in the same thread one message above [1] of defining a `foo` var twice with the type in question (i.e. `time_t`, `off_t`, etc.) and a supposed type, e.g.:

```
extern time_t foo;
extern long int foo;
```

If it compiles, then `time_t` is `long int`. Otherwise it throws an error like this (tested with both clang and gcc, both behave the same on both Linux and OpenBSD, so this technique works well, but a bit cumbersome):

gcc (5.4.0) & clang on Linux:
```
test.c:7:22: error: conflicting types for ‘foo’
 extern long long int foo;
                      ^
test.c:6:15: note: previous declaration of ‘foo’ was here
 extern time_t foo;
               ^


test.c:7:22: error: redeclaration of 'foo' with a different type: 'long long' vs 'time_t' (aka 'long')
extern long long int foo;
                     ^
test.c:6:15: note: previous declaration is here
extern time_t foo;
              ^
1 error generated.
```

gcc (4.2.1) & clang on OpenBSD (after changing `long long int` for `long int`):
```
test.c:7: error: conflicting types for 'foo'
test.c:6: error: previous declaration of 'foo' was here


test.c:7:17: error: redeclaration of 'foo' with a different type: 'long' vs 'time_t' (aka 'long long')
extern long int foo;
                ^
test.c:6:15: note: previous declaration is here
extern time_t foo;
              ^
1 error generated.
```

[1] https://lists.gnu.org/archive/html/autoconf/2020-10/msg00042.html

**Comment by anatoli26** [comment]
And there's a `__builtin_types_compatible_p` built-in function available at least since gcc 3.1.1 (2002) [1] and clang 1.0 (2009) [2] which could be used similarly to `_Generic`:

```
#define TIME_T_FMT (__builtin_types_compatible_p(time_t, long long int) ? "%lld" : \
                    __builtin_types_compatible_p(time_t, long int) ? "%ld" : "%d")
```

[1] https://gcc.gnu.org/onlinedocs/gcc-3.1.1/gcc/Other-Builtins.html
[2] https://releases.llvm.org/1.0/docs/ReleaseNotes.html

**Comment by elliefm** [comment]
The `__builtin_types_compatible_p` version doesn't seem to work for me, I think because in the particular context where we try to use TIME_T_FMT (implicitly concatenating together, at compile time, a series of adjacent string literals) we want a compile-time string literal, but instead it's now an expression that only evaluates to a string literal at run-time.

I don't know much about C11; I haven't bothered to spend any time learning it yet, since we've only just recently been able to rely on C99 constructs being available.  I don't know whether `_Generic` is a compile-time or run-time thing, though if it's a run-time thing it seems like it might suffer the same problem here as `__builtin_types_compatible_p` does.

The redefined-type test looks promising though!  The only types that are troublesome (so far) are `off_t` and `time_t`, and we know that each of those is either `long int` or `long long int`, so it wouldn't be too burdensome to just use AC_COMPILE_IFELSE checks instead of comparing the sizes.

**Comment by elliefm** [comment]
I'm having a play with building a macro that uses the redefined-type check to figure out what to use.  Here's where I'm at so far: https://github.com/cyrusimap/cyrus-imapd/compare/master...elliefm:v33/3128-time-offset-format-detection

**Comment by anatoli26** [comment]
@elliefm, my understanding with respect to `__builtin_types_compatible_p` is that it's an internal compiler function that is resolved during the compilation, though the result is available at run-time, but despite this, it should be optimized to a build-time value during compilation.

In any case, all these functions could be used to run tests during `./configure` stage and obtain the correct format specifier, e.g. something like this (and there the `#define TIME_T_FMT` is working fine for me too):

```
#include <stdio.h>
#include <time.h>

#define TIME_T_FMT (__builtin_types_compatible_p(time_t, long int) ? "%ld" : \
                    __builtin_types_compatible_p(time_t, long long int) ? "%lld" : "unknown")

int main(int argc, char **argv) {
    time_t s;
    if (__builtin_types_compatible_p(time_t, long int)) {
        puts("long int");
    } else if (__builtin_types_compatible_p(time_t, long long int)) {
        puts("long long int");
    }

    puts(TIME_T_FMT);
    return (0);
};
```

Anyway, as you said, the redefined-type test could be the best solution as its logic is based on the same judgement the compiler does when determining printf format specifier mismatches.

**Comment by anatoli26** [comment]
> The `__builtin_types_compatible_p` version doesn't seem to work for me, I think because in the particular context where we try to use TIME_T_FMT (implicitly concatenating together, at compile time, a series of adjacent string literals) we want a compile-time string literal, but instead it's now an expression that only evaluates to a string literal at run-time.

@elliefm, I've misread your comment, got it now about the implicit concat of the `#define TIME_T_FMT`, please disregard my 1st paragraph in the post above, though the rest is ok (i.e. to use this func in a `configure.am` test).

**Comment by elliefm** [comment]
Fixed on master and 3.2 by #3262

**Comment by SamuelMarks** [comment]
FreeSwitch has a pretty exhaustive way of handling these https://github.com/signalwire/freeswitch/blob/master/src/include/switch_platform.h
